### ISSUE TITLE
Introduce codestyle set

### ISSUE THREAD

**Issue by Tapchicoma** [pinterest/ktlint#548]
## Proposal

Ktlint should introduce codestyle-set concept. Codestyle-set will contain set of rules, that would be used by ktlint on linting/formatting the code.

Such approach:
- reduces the level of user confusion (see #533, #527, #544) by allowing changes in defaults to be more conservative
- still easy to consume
- is flexible. User can change codestyle-set to another or custom one. Custom codestyle-set can extend existing codestyle-set, rules can be reused in different codestyle-sets.
- allows without much effort introduce support for other Kotlin code styles in the future (like Google Kotlin codestyle).

## Implementation details

Each codestyle-set should have a unique name and contain list of rule ids. Codestyle-set could extend another codestyle-set by adding (removing (?)) rules to the parent one.

### Built-in

Ktlint should provide built-in codestyle-sets:
- "official" (default) provides official Kotlin styleguide and should works the same way as IDEA Kotlin code style formatting
- "official-experimental" - adds experimental rules on top of "official" codestyle-set
- "android" - provides official Android Kotlin styleguide support
- "android-experimental" - adds experimental rules on top of "android" codestyle-set

### `.Editorconfig`

User should be able to define and use custom codestyle-set, using `.editorconfig` file:
```editorconfig
[*.{kt,kts}]
codestyle-set-define=my-kotlin-style-name:final-newline,chain-wrapping,no-unit-return
codestyle-set-define=my-kotlin-style-name-extended:my-kotlin-style-name:no-unused-imports,comma-spacing
codestyle-set=my-kotlin-style-name
```

where `codestyle-set-define` has following format:
```editorconfig
codestyle-set-define=<set name>:<optional parent set name>:<list of rule ids separated by comma>
```

`.Editorconfig` can contain several `codestyle-set` entries, but only last one `codestyle-set` entry will be used by Ktlint.
Nested `.editorconfig` files could override `codestyle-set` entry used in parent `.editorconfig`.

### Command line

Usage of codestyle-set in command line looks similar to how it is done in `.editorconfig` file:
```bash
$ ktlint --codestyle-set-define=my-kotlin-style-name:official:final-newline,chain-wrapping,no-unit-return --codestyle-set=my-kotlin-style-name
```

## Deprecations

`--android` CLI flag should be replace by "android" codestyle-set

**Comment by paul-dingemans** [comment]
> ## Deprecations
> `--android` CLI flag should be replace by "android" codestyle-set

With #1488 the `.editorconfig` property `ktlint_code_style` is introduced making this command line flag redundant. This flag might be removed in a future release.

I see no reason for further implementation of "flexible" code style rules. As no one else has attempted to introduce this in the last 3 years, the issue is closed.

**Comment by idntfy** [comment]
@paul-dingemans as it was described in the ticket from which this ticket is originated, .editorconfig is not a scalable option, this is why the flag is requested. You can read more details here: https://github.com/pinterest/ktlint/issues/533

**Comment by paul-dingemans** [comment]
> @paul-dingemans as it was described in the ticket from which this ticket is originated, .editorconfig is not a scalable option, this is why the flag is requested. You can read more details here: https://github.com/pinterest/ktlint/issues/533

In last couple of years, the focus is shifting more and more towards `.editorconfig` settings. From a maintenance perspective te shere varity of CLI and API flags is becoming unmaintainable. So we are going to phaseout more and more of those flag in favour of `.editorconfig` properties. 

> The problem with ktlint is that it is android-centric and doesn't use official code style.

I don't believe this is true. The default code style if the official Kotlin codestyle. If you want to use `android` codestyle this has to be specified on the command line or a special flag. When not specifying the `ktlint_code_style` property in `.editorconfig` then the offical Kotlin code style is used.

In case your company does not use `.editorconfig` at all, you might want to look into the CLI/API flag that allows to specify a additional `.editorconfig` file. This could be the same file for your entire company and as of that can be maintained in a single place. But please do note that it will take precedence on any `.editorconfig` file that is specified in any of the subprojects.
